File: objc/Project7/Project7/SearchViewController.m
//
//  SearchViewController.m
//  Project7
//
//  Created by Jinwoo Kim on 2/11/21.
//

#import "SearchViewController.h"

@implementation SearchViewController

- (void)setup {
    self.allCities = [@[] mutableCopy];
    self.matchingCities = @[];
}

- (void)viewDidLoad {
    [self setup];
    [super viewDidLoad];
    
    NSURL *capitals = [NSBundle.mainBundle URLForResource:@"capitals" withExtension:@"json"];
    if (capitals == nil) return;
    NSData *contents = [NSData dataWithContentsOfURL:capitals];
    if (contents == nil) return;
    
    NSArray<NSDictionary *> *cities = [NSJSONSerialization JSONObjectWithData:contents options:0 error:nil];
    
    [cities enumerateObjectsUsingBlock:^(NSDictionary * _Nonnull city, NSUInteger idx, BOOL * _Nonnull stop) {
        CLLocationCoordinate2D capitals = CLLocationCoordinate2DMake([(NSNumber *)city[@"lat"] doubleValue],
                                                                     [(NSNumber *)city[@"lon"] doubleValue]);
        
        City *newCity = [[City alloc] initWithName:(NSString *)city[@"name"]
                                           country:(NSString *)city[@"country"]
                                       coordinates:capitals];
        
        [self.allCities addObject:newCity];
    }];
    NSLog(@"%lu", cities.count);
    
//    NSSortDescriptor *descriptor = [NSSortDescriptor sortDescriptorWithKey:@"name" ascending:YES];
//    [self.allCities sortUsingDescriptors:@[descriptor]];
    self.allCities = [[self.allCities sortedArrayUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
        City *city1 = (City *)obj1;
        City *city2 = (City *)obj2;
        if ((city1 == nil) || (city2 == nil)) return NSOrderedSame;
        return [city1.name compare:city2.name];
    }] mutableCopy];
}

#pragma UITableViewDataSource, UITableViewDelegate

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.matchingCities.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    City *city = self.matchingCities[indexPath.row];
    cell.textLabel.text = city.formattedName;
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    ViewController *map = (ViewController *)[self.mainTabBarController.viewControllers firstObject];
    if (map == nil) return;
    
    City *city = self.matchingCities[indexPath.row];
    [map focusOn:city];
    
    self.mainTabBarController.selectedIndex = 0;
}

#pragma UISearchResultsUpdating

- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
    NSString *search = searchController.searchBar.text;
    if (search == nil) return;
    
    if (([search isEqualToString:@""]) || (search == nil)) {
        // no text - return all cities
        self.matchingCities = self.allCities;
    } else {
        // run search
        NSPredicate *predicate = [NSPredicate predicateWithBlock:^BOOL(id  _Nullable evaluatedObject, NSDictionary<NSString *,id> * _Nullable bindings) {
            City *city = (City *)evaluatedObject;
            if (city == nil) return NO;
            return [city matchesWithText:search];
        }];
        self.matchingCities = [self.allCities filteredArrayUsingPredicate:predicate];
    }
    
    [self.tableView reloadData];
}

@end

File: swift/Project7/Project7/SearchViewController.swift
//
//  SearchViewController.swift
//  Project7
//
//  Created by Jinwoo Kim on 2/11/21.
//

import UIKit
import CoreLocation

class SearchViewController: UIViewController, UISearchResultsUpdating, UITableViewDataSource, UITableViewDelegate {
    @IBOutlet weak var tableView: UITableView!
    weak var mainTabBarController: UITabBarController!
    var allCities = [City]()
    var matchingCities = [City]()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        guard let capitals = Bundle.main.url(forResource: "capitals", withExtension: "json") else { return }
        guard let contents = try? Data(contentsOf: capitals) else { return }
        
        let cities = JSON(contents).arrayValue
        
        for city in cities {
            let coords = CLLocationCoordinate2D(latitude: city["lat"].doubleValue, longitude: city["lon"].doubleValue)
            let newCity = City(name: city["name"].stringValue, country: city["country"].stringValue, coordinates: coords)
            allCities.append(newCity)
        }
        
        allCities.sort()
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return matchingCities.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        let city = matchingCities[indexPath.row]
        cell.textLabel?.text = city.formattedName
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        guard let map = mainTabBarController.viewControllers?.first as? ViewController else {
            return
        }
        
        let city = matchingCities[indexPath.row]
        map.focus(on: city)
        
        mainTabBarController.selectedIndex = 0;
    }
    
    func updateSearchResults(for searchController: UISearchController) {
        guard let search = searchController.searchBar.text else { return }
        
        if search.isEmpty {
            // no text - return all cities
            matchingCities = allCities
        } else {
            // run search
            matchingCities = allCities.filter {
                $0.matches(search)
            }
        }
        
        tableView.reloadData()
    }
}
